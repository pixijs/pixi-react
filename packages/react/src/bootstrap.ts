import { configurePixiReact } from '@pixi/react-modular';
import { configurePixiReactFiber, configurePixiReactHostConfig } from '@pixi/react-fiber';
import { configurePixiReactComponents, configurePixiReactStage, configurePixiReactRenderAPI } from '@pixi/react-components';
import type { PixiReactHostConfig, PixiReactReconciler } from '@pixi/react-fiber';
import type { PixiReactContainer, ReactStageComponent } from '@pixi/react-components';

const { PixiComponent, PixiReactFiber, Stage, createRoot, render, unmountComponentAtNode } = configurePixiReact<
ReactStageComponent,
PixiReactContainer,
PixiReactContainer,
PixiReactHostConfig<PixiReactContainer, PixiReactContainer>,
PixiReactReconciler<PixiReactContainer, PixiReactContainer>
>({
    configurePixiReactHostConfig,
    configurePixiReactFiber,
    configurePixiReactComponents,
    configurePixiReactStage,
    configurePixiReactRenderAPI,
});

export { PixiComponent, PixiReactFiber, Stage, createRoot, render, unmountComponentAtNode };
