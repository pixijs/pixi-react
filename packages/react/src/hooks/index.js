export * from './useTick';
export * from './useApp';
