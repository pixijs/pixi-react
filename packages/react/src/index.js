import { TYPES } from './utils/element';

/**
 * -------------------------------------------
 * Public API
 * -------------------------------------------
 */

export * from './exports';

export const BitmapText = TYPES.BitmapText;
export const Container = TYPES.Container;
export const Graphics = TYPES.Graphics;
export const NineSlicePlane = TYPES.NineSlicePlane;
export const ParticleContainer = TYPES.ParticleContainer;
export const Sprite = TYPES.Sprite;
export const AnimatedSprite = TYPES.AnimatedSprite;
export const Text = TYPES.Text;
export const TilingSprite = TYPES.TilingSprite;
export const SimpleMesh = TYPES.SimpleMesh;
export const SimpleRope = TYPES.SimpleRope;
