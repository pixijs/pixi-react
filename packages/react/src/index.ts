import {
    TYPES,
    BitmapText,
    Container,
    Graphics,
    NineSlicePlane,
    ParticleContainer,
    Sprite,
    AnimatedSprite,
    Text,
    TilingSprite,
    SimpleMesh,
    SimpleRope,
    AppProvider,
    AppConsumer,
    AppContext,
    applyDefaultProps,
    eventHandlers,
    useTick,
    useApp,
    withPixiApp,
    withFilters,
} from '@pixi/react-components';
import {
    PixiComponent,
    PixiReactFiber,
    Stage,
    createRoot,
    render,
    unmountComponentAtNode,
} from './bootstrap';

export {
    PixiComponent,
    PixiReactFiber,
    TYPES,
    BitmapText,
    Container,
    Graphics,
    NineSlicePlane,
    ParticleContainer,
    Sprite,
    AnimatedSprite,
    Text,
    TilingSprite,
    SimpleMesh,
    SimpleRope,
    createRoot,
    render,
    unmountComponentAtNode,
    Stage,
    AppProvider,
    AppConsumer,
    AppContext,
    useTick,
    useApp,
    applyDefaultProps,
    withPixiApp,
    withFilters,
    eventHandlers,
};
