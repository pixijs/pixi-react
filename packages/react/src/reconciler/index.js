import Reconciler from 'react-reconciler';
import pkg from '../../../../package.json';
import hostconfig from './hostconfig';

export const PixiFiber = Reconciler(hostconfig);

export const VERSION = pkg.version;
export const REACT_DOM_VERSION = pkg.dependencies['react-dom'].replace(/[^0-9.]/g, '');
export const PACKAGE_NAME = pkg.name;

PixiFiber.injectIntoDevTools({
    bundleType: process.env.NODE_ENV !== 'production' ? 1 : 0,
    version: REACT_DOM_VERSION,
    rendererPackageName: PACKAGE_NAME,
    findHostInstanceByFiber: PixiFiber.findHostInstance,
});
