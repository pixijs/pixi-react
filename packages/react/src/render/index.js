import { Container } from '@pixi/display';
import invariant from '../utils/invariant';
import { PixiFiber } from '../reconciler';

// cache both root PixiFiber containers and React roots
export const roots = new Map();

/**
 * @param {Container} container
 */
function unmountComponent(container)
{
    invariant(
        Container.prototype.isPrototypeOf(container),
        'Invalid argument `container`, expected instance of `PIXI.Container`.'
    );

    if (roots.has(container))
    {
        const { pixiFiberContainer } = roots.get(container);

        // unmount component
        PixiFiber.updateContainer(null, pixiFiberContainer, undefined, () =>
        {
            roots.delete(container);
        });
    }
}

/**
 * Custom Renderer with react 18 API
 * Use this without React-DOM
 *
 * @param {Container} container
 * @returns {{ render: Function, unmount: Function}}
 */
export function createRoot(container)
{
    invariant(
        Container.prototype.isPrototypeOf(container),
        'Invalid argument `container`, expected instance of `PIXI.Container`.'
    );

    let root = roots.get(container);

    invariant(!root, 'Pixi React: createRoot should only be called once');

    if (!root)
    {
        const pixiFiberContainer = PixiFiber.createContainer(container);

        const reactRoot = {
            render(element)
            {
                // schedules a top level update
                PixiFiber.updateContainer(
                    element,
                    pixiFiberContainer,
                    undefined
                );

                return PixiFiber.getPublicRootInstance(pixiFiberContainer);
            },
            unmount()
            {
                unmountComponent(container);
                roots.delete(container);
            },
        };

        root = { pixiFiberContainer, reactRoot };
        roots.set(container, root);
    }

    return root.reactRoot;
}

/**
 * Custom Renderer
 * Use this without React-DOM
 *
 * @deprecated use createRoot instead
 *
 * @param {React.ReactNode} element
 * @param {Container} container (i.e. the Stage)
 * @param {Function} callback
 */
export function render(element, container, callback)
{
    console.warn(
        'Pixi React Deprecation Warning: render is deprecated, use createRoot instead'
    );

    if (callback !== undefined)
    {
        console.warn(
            'Pixi React Deprecation Warning: render callback no longer exists in React 18'
        );
    }

    let reactRoot;

    if (roots.has(container))
    {
        ({ reactRoot } = roots.get(container));
    }
    else
    {
        reactRoot = createRoot(container);
    }

    return reactRoot.render(element);
}

/**
 * @deprecated use root.unmount() instead
 * @param {Container} container
 */
export function unmountComponentAtNode(container)
{
    unmountComponent(container);
}
