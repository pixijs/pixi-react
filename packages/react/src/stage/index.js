import React from 'react';
import { Application } from '@pixi/app';
import { Ticker } from '@pixi/ticker';
import PropTypes from 'prop-types';
import invariant from '../utils/invariant';
import { PROPS_DISPLAY_OBJECT } from '../utils/props';
import { PixiFiber } from '../reconciler';
import { AppProvider } from './provider';

const noop = () => {};

/**
 * -------------------------------------------
 * Stage React Component (use this in react-dom)
 *
 * @usage
 *
 * const App = () => (
 *   <Stage
 *     width={500}
 *     height={500}
 *     options={ backgroundColor: 0xff0000 }
 *     onMount={( renderer, canvas ) => {
 *       console.log('PIXI renderer: ', renderer)
 *       console.log('Canvas element: ', canvas)
 *     }}>
 * );
 *
 * -------------------------------------------
 */

const propTypes = {
    // dimensions
    width: PropTypes.number,
    height: PropTypes.number,

    // will return renderer
    onMount: PropTypes.func,
    onUnmount: PropTypes.func,

    // run ticker at start?
    raf: PropTypes.bool,

    // render component on component lifecycle changes?
    renderOnComponentChange: PropTypes.bool,

    children: PropTypes.node,

    // PIXI options, see https://pixijs.download/dev/docs/PIXI.Application.html
    options: PropTypes.shape({
        autoStart: PropTypes.bool,
        width: PropTypes.number,
        height: PropTypes.number,
        useContextAlpha: PropTypes.bool,
        backgroundAlpha: PropTypes.number,
        autoDensity: PropTypes.bool,
        antialias: PropTypes.bool,
        preserveDrawingBuffer: PropTypes.bool,
        resolution: PropTypes.number,
        forceCanvas: PropTypes.bool,
        backgroundColor: PropTypes.number,
        clearBeforeRender: PropTypes.bool,
        powerPreference: PropTypes.string,
        sharedTicker: PropTypes.bool,
        sharedLoader: PropTypes.bool,

        // resizeTo needs to be a window or HTMLElement
        resizeTo: (props, propName) =>
        {
            const el = props[propName];

            el
                && invariant(
                    el === window || el instanceof HTMLElement,
                    `Invalid prop \`resizeTo\` of type ${typeof el}, expect \`window\` or an \`HTMLElement\`.`
                );
        },

        // view is optional, use if provided
        view: (props, propName, componentName) =>
        {
            const el = props[propName];

            el
                && invariant(
                    el instanceof HTMLCanvasElement,
                    `Invalid prop \`view\` of type ${typeof el}, supplied to ${componentName}, expected \`<canvas> Element\``
                );
        },
    }),
};

const defaultProps = {
    width: 800,
    height: 600,
    onMount: noop,
    onUnmount: noop,
    raf: true,
    renderOnComponentChange: true,
};

export function getCanvasProps(props)
{
    const reserved = [
        ...Object.keys(propTypes),
        ...Object.keys(PROPS_DISPLAY_OBJECT),
    ];

    return Object.keys(props)
        .filter((p) => !reserved.includes(p))
        .reduce((all, prop) => ({ ...all, [prop]: props[prop] }), {});
}

class Stage extends React.Component
{
    _canvas = null;
    _mediaQuery = null;
    _ticker = null;
    _needsUpdate = true;
    app = null;

    componentDidMount()
    {
        const {
            onMount,
            width,
            height,
            options,
            raf,
            renderOnComponentChange,
        } = this.props;

        this.app = new Application({
            width,
            height,
            view: this._canvas,
            ...options,
            autoDensity: options?.autoDensity !== false,
        });

        if (process.env.NODE_ENV === 'development')
        {
            // workaround for React 18 Strict Mode unmount causing
            // webgl canvas context to be lost
            if (this.app.renderer.context?.extensions)
            {
                this.app.renderer.context.extensions.loseContext = null;
            }
        }

        this.app.ticker.autoStart = false;
        this.app.ticker[raf ? 'start' : 'stop']();

        this.app.stage.__reactpixi = { root: this.app.stage };
        this.mountNode = PixiFiber.createContainer(this.app.stage);
        PixiFiber.updateContainer(this.getChildren(), this.mountNode, this);

        onMount(this.app);

        // update size on media query resolution change?
        // only if autoDensity = true
        if (
            options?.autoDensity
            && window.matchMedia
            && options?.resolution === undefined
        )
        {
            this._mediaQuery = window.matchMedia(
                `(-webkit-min-device-pixel-ratio: 1.3), (min-resolution: 120dpi)`
            );
            this._mediaQuery.addListener(this.updateSize);
        }

        // listen for reconciler changes
        if (renderOnComponentChange && !raf)
        {
            this._ticker = new Ticker();
            this._ticker.autoStart = true;
            this._ticker.add(this.renderStage);
            this.app.stage.on(
                '__REACT_PIXI_REQUEST_RENDER__',
                this.needsRenderUpdate
            );
        }

        this.updateSize();
        this.renderStage();
    }

    componentDidUpdate(prevProps, prevState, prevContext)
    {
        const { width, height, raf, renderOnComponentChange, options }
            = this.props;

        // update resolution
        if (
            options?.resolution !== undefined
            && prevProps?.options.resolution !== options?.resolution
        )
        {
            this.app.renderer.resolution = options.resolution;
            this.resetInteractionManager();
        }

        // update size
        if (
            prevProps.height !== height
            || prevProps.width !== width
            || prevProps.options?.resolution !== options?.resolution
        )
        {
            this.updateSize();
        }

        // handle raf change
        if (prevProps.raf !== raf)
        {
            this.app.ticker[raf ? 'start' : 'stop']();
        }

        // flush fiber
        PixiFiber.updateContainer(this.getChildren(), this.mountNode, this);

        if (
            prevProps.width !== width
            || prevProps.height !== height
            || prevProps.raf !== raf
            || prevProps.renderOnComponentChange !== renderOnComponentChange
            || prevProps.options !== options
        )
        {
            this._needsUpdate = true;
            this.renderStage();
        }
    }

    updateSize = () =>
    {
        const { width, height, options } = this.props;

        if (!options?.resolution)
        {
            this.app.renderer.resolution = window.devicePixelRatio;
            this.resetInteractionManager();
        }

        this.app.renderer.resize(width, height);
    };

    needsRenderUpdate = () =>
    {
        this._needsUpdate = true;
    };

    renderStage = () =>
    {
        const { renderOnComponentChange, raf } = this.props;

        if (!raf && renderOnComponentChange && this._needsUpdate)
        {
            this._needsUpdate = false;
            this.app.renderer.render(this.app.stage);
        }
    };

    // provide support for Pixi v6 still
    resetInteractionManager()
    {
        // `interaction` property is absent in Pixi v7 and in v6 if user has installed Federated Events API plugin.
        // https://api.pixijs.io/@pixi/events.html
        // in v7 however, there's a stub object which displays a deprecation warning, so also check the resolution property:
        const { interaction: maybeInteraction } = this.app.renderer.plugins;

        if (maybeInteraction?.resolution)
        {
            maybeInteraction.resolution = this.app.renderer.resolution;
        }
    }

    getChildren()
    {
        const { children } = this.props;

        return <AppProvider value={this.app}>{children}</AppProvider>;
    }

    componentDidCatch(error, errorInfo)
    {
        console.error(`Error occurred in \`Stage\`.`);
        console.error(error);
        console.error(errorInfo);
    }

    componentWillUnmount()
    {
        this.props.onUnmount(this.app);

        if (this._ticker)
        {
            this._ticker.remove(this.renderStage);
            this._ticker.destroy();
        }

        this.app.stage.off(
            '__REACT_PIXI_REQUEST_RENDER__',
            this.needsRenderUpdate
        );

        PixiFiber.updateContainer(null, this.mountNode, this);

        if (this._mediaQuery)
        {
            this._mediaQuery.removeListener(this.updateSize);
            this._mediaQuery = null;
        }

        this.app.destroy();
    }

    render()
    {
        const { options } = this.props;

        if (options && options.view)
        {
            invariant(
                options.view instanceof HTMLCanvasElement,
                'options.view needs to be a `HTMLCanvasElement`'
            );

            return null;
        }

        return (
            <canvas
                {...getCanvasProps(this.props)}
                ref={(c) => (this._canvas = c)}
            />
        );
    }
}

Stage.propTypes = propTypes;
Stage.defaultProps = defaultProps;

export default Stage;
