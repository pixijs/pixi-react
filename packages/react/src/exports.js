import { PixiComponent, TYPES } from './utils/element';
import { createRoot, render, unmountComponentAtNode } from './render';
import Stage from './stage';
import { PixiFiber } from './reconciler';
import { Context as AppContext, AppProvider, AppConsumer, withPixiApp } from './stage/provider';
import { useTick, useApp } from './hooks';
import { withFilters } from './hoc';
import { eventHandlers } from './utils/pixi';
import { applyDefaultProps } from './utils/props';

/**
 * -------------------------------------------
 * Public API
 * -------------------------------------------
 */

export {
    createRoot,
    render,
    unmountComponentAtNode,
    Stage,
    withPixiApp,
    PixiComponent,
    PixiFiber,
    AppProvider,
    AppConsumer,
    AppContext,
    useTick,
    useApp,
    withFilters,
    applyDefaultProps,
    eventHandlers,
    TYPES,
};
