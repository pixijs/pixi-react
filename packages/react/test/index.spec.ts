import * as index from '../src/index';

describe('index', () =>
{
    test('export modules for pixi-react', () =>
    {
        expect(index).toMatchSnapshot();
    });
});
