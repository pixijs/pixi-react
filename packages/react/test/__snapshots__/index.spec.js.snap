// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`index export modules for index 1`] = `
{
  "AnimatedSprite": "AnimatedSprite",
  "AppConsumer": {
    "$$typeof": Symbol(react.context),
    "_context": {
      "$$typeof": Symbol(react.context),
      "Consumer": [Circular],
      "Provider": {
        "$$typeof": Symbol(react.provider),
        "_context": [Circular],
      },
      "_currentRenderer": null,
      "_currentRenderer2": null,
      "_currentValue": null,
      "_currentValue2": null,
      "_defaultValue": null,
      "_globalName": null,
      "_threadCount": 0,
    },
  },
  "AppContext": {
    "$$typeof": Symbol(react.context),
    "Consumer": {
      "$$typeof": Symbol(react.context),
      "_context": [Circular],
    },
    "Provider": {
      "$$typeof": Symbol(react.provider),
      "_context": [Circular],
    },
    "_currentRenderer": null,
    "_currentRenderer2": null,
    "_currentValue": null,
    "_currentValue2": null,
    "_defaultValue": null,
    "_globalName": null,
    "_threadCount": 0,
  },
  "AppProvider": {
    "$$typeof": Symbol(react.provider),
    "_context": {
      "$$typeof": Symbol(react.context),
      "Consumer": {
        "$$typeof": Symbol(react.context),
        "_context": [Circular],
      },
      "Provider": [Circular],
      "_currentRenderer": null,
      "_currentRenderer2": null,
      "_currentValue": null,
      "_currentValue2": null,
      "_defaultValue": null,
      "_globalName": null,
      "_threadCount": 0,
    },
  },
  "BitmapText": "BitmapText",
  "Container": "Container",
  "Graphics": "Graphics",
  "NineSlicePlane": "NineSlicePlane",
  "ParticleContainer": "ParticleContainer",
  "PixiComponent": [Function],
  "PixiFiber": {
    "attemptContinuousHydration": [Function],
    "attemptDiscreteHydration": [Function],
    "attemptHydrationAtCurrentPriority": [Function],
    "attemptSynchronousHydration": [Function],
    "batchedUpdates": [Function],
    "createComponentSelector": [Function],
    "createContainer": [Function],
    "createHasPseudoClassSelector": [Function],
    "createHydrationContainer": [Function],
    "createPortal": [Function],
    "createRoleSelector": [Function],
    "createTestNameSelector": [Function],
    "createTextSelector": [Function],
    "deferredUpdates": [Function],
    "discreteUpdates": [Function],
    "findAllNodes": [Function],
    "findBoundingRects": [Function],
    "findHostInstance": [Function],
    "findHostInstanceWithNoPortals": [Function],
    "findHostInstanceWithWarning": [Function],
    "flushControlled": [Function],
    "flushPassiveEffects": [Function],
    "flushSync": [Function],
    "focusWithin": [Function],
    "getCurrentUpdatePriority": [Function],
    "getFindAllNodesFailureDescription": [Function],
    "getPublicRootInstance": [Function],
    "injectIntoDevTools": [Function],
    "isAlreadyRendering": [Function],
    "observeVisibleRects": [Function],
    "registerMutableSourceForHydration": [Function],
    "runWithPriority": [Function],
    "shouldError": [Function],
    "shouldSuspend": [Function],
    "updateContainer": [Function],
  },
  "SimpleMesh": "SimpleMesh",
  "SimpleRope": "SimpleRope",
  "Sprite": "Sprite",
  "Stage": [Function],
  "TYPES": {
    "AnimatedSprite": "AnimatedSprite",
    "BitmapText": "BitmapText",
    "Container": "Container",
    "Graphics": "Graphics",
    "NineSlicePlane": "NineSlicePlane",
    "ParticleContainer": "ParticleContainer",
    "SimpleMesh": "SimpleMesh",
    "SimpleRope": "SimpleRope",
    "Sprite": "Sprite",
    "Text": "Text",
    "TilingSprite": "TilingSprite",
  },
  "Text": "Text",
  "TilingSprite": "TilingSprite",
  "applyDefaultProps": [Function],
  "createRoot": [Function],
  "eventHandlers": [
    "click",
    "mousedown",
    "mousemove",
    "mouseout",
    "mouseover",
    "mouseup",
    "mouseupoutside",
    "tap",
    "touchstart",
    "touchmove",
    "touchend",
    "touchendoutside",
    "pointercancel",
    "pointerout",
    "pointerover",
    "pointertap",
    "pointerdown",
    "pointerup",
    "pointerupoutside",
    "pointermove",
    "rightclick",
    "rightdown",
    "rightup",
    "rightupoutside",
    "touchcancel",
    "touchendoutside",
    "touchmove",
    "touchstart",
  ],
  "render": [Function],
  "unmountComponentAtNode": [Function],
  "useApp": [Function],
  "useTick": [Function],
  "withFilters": [Function],
  "withPixiApp": [Function],
}
`;
